### PR TITLE
Load tweaks necessary to load the Rabinowitz data

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

This includes changes necessary to resolve issues encountered when attempting to load the Rabinowitz data.  To be precise, one study (`MN_exp027e_KR_infusion`) contained 2 accucor files of MSRun data.  The 2 files contained the same samples, and mostly had no compound overlap, except 1: `proline`.  When the second file load tried to do a `get_or_create` for the peak group for proline, it tried to create a new record (because the peak annotation file linked in the peak group record differed), but it had associated the peakgroup with an existing `MSRunSample`, because all of its parameters were the same (same sequence, same sample, same polarity (positive).  Both sets of MSRuns only differed by scan range.  One was high mz and one was low mz.  So to solve this, I added `mz_min` and `mz_max` to `MSRunSample` and updated the loading code to accept those inputs.

## Affected Issues/Pull Requests

- Partially addresses https://github.com/PrincetonUniversity/tracebase-rabinowitz-data/issues/80
- Partially addresses https://github.com/PrincetonUniversity/tracebase-rabinowitz-data/issues/81
- Partially addresses https://github.com/PrincetonUniversity/tracebase-rabinowitz-data/issues/82
- Merges into PR #831

## Review Notes

See comments in-line.

## Checklist

This pull request will be merged once the following requirements are met.  The
author and/or reviewers should uncheck any unmet requirements:

- Review requirements
  - Minimum approvals: 1 
  - No changes requested
  - All blocking issues resolved by reviewers
  - Specific reviewers: @__add_username_here__
    
  - Review period: 2 days 
- Associated issue/pull request requirements:
  
  - [x] All requirements in affected issues marked "resolved" are satisfied
  - [x] All required pull requests are merged *(or none)*
- Basic requirements
  
  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All conflicts resolved
- Overhead requirements
  
  - [x] [New/changed method tests implemented/updated *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [x] [Updated *Unreleased* section of `changelog.md` *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md)
  - [x] [Migrations created & committed *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)

### CODE REVIEW

File: DataRepo/models/msrun_sample.py [hunk 1]
```diff
@@ -119,12 +143,20 @@ class Meta:
...
+                    "mz max": self.mz_max,
+                    "RAW file": self.ms_raw_file,
+                    "mzXML": self.ms_data_file,
+                }.items()
                 if i is not None
             ]
         )
-        return f"MS run of sample {self.sample} in {self.msrun_sequence} ({details})"
+        if details == "":
+            details = "using default parameters"
```
Comment (hepcat72) [inline DataRepo/models/msrun_sample.py:158]: C. This was to prevent an ugly occurrence of `()` in the return string.

File: DataRepo/models/msrun_sample.py [hunk 2]
```diff
@@ -119,12 +143,20 @@ class Meta:
...
             [
-                i
-                for i in [self.polarity, self.ms_raw_file, self.ms_data_file]
+                f"{k}: {i}"
+                for k, i in {
+                    "polarity": self.polarity,
+                    "mz min": self.mz_min,
+                    "mz max": self.mz_max,
+                    "RAW file": self.ms_raw_file,
+                    "mzXML": self.ms_data_file,
```
Comment (hepcat72) [inline DataRepo/models/msrun_sample.py:152]: C. Changed to a dict to be able to display leaders, because otherwise, you end up potentially seeing "...unknown unknown..."

File: DataRepo/models/msrun_sequence.py [hunk 3]
```diff
@@ -73,6 +73,6 @@ class Meta:
 
     def __str__(self):
         return str(
-            f"MS run sequence using instrument {self.instrument} with {self.lc_method.name} by {self.researcher} on "
-            f"{self.date}"
+            f"MS run sequence on instrument [{self.instrument}] with LC protocol [{self.lc_method.name}], operated by "
+            f"{self.researcher} on {self.date}"
```
Comment (hepcat72) [inline DataRepo/models/msrun_sequence.py:77]: C. Makes occurrences (e.g. in errors) easier to interpret.

File: DataRepo/models/utilities.py [hunk 4]
```diff
@@ -277,20 +277,24 @@ def check_for_inconsistencies(rec, rec_dict, rownum=None, sheet=None, file=None)
...
+    if len(differences.keys()) > 0:
+        conflicting_value_errors.append(
+            ConflictingValueError(
+                rec,
+                differences,
+                rownum=rownum,
+                sheet=sheet,
+                file=file,
             )
+        )
```
Comment (hepcat72) [inline DataRepo/models/utilities.py:297]: C. I made this change because it was bugging me that it would only report the first difference it encounters, which was hampering my efforts on this issue.  Now it shows all the differences.

File: DataRepo/models/utilities.py [hunk 5]
```diff
@@ -363,23 +367,22 @@ def handle_load_db_errors(
...
                         rec, rec_dict, rownum=rownum, sheet=sheet, file=file
                     )
-                    if conflicts_list is not None:
-                        conflicts_list.extend(errs)
-                        return True
-                    elif aes:
-                        for err in errs:
-                            aes.buffer_error(err)
-                        return True
+                    if len(errs) > 0:
```
Comment (hepcat72) [inline DataRepo/models/utilities.py:370]: C. Encountered this bug in the course of this implementation.  Errors weren't being raised because nothing was getting appended to `conflicts_list`.

File: DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_glucose_conflicting.xlsx
Comment (hepcat72) [inline DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_glucose_conflicting.xlsx:1]: C. This file was edited because the test that uses it was originally intended to generate a `ConflictingValueErrors` exception, but there's a new exception called `AmbiguousMSRuns` that is intended to supply useful information of an indirect edge case (that arises from `PeakGroup` when the MSRun data is the same between 2 separate runs (e.g. `lowMz`/`highMz` runs).  So this test had started raising that error instead because the only difference was the `ArchiveFile` foreign key about the peak annotation file.  So I needed a difference between the runs other than the peak annotation file.  So I changed the formula for `glucose` in the file.

File: DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_glucose_mzxmls/BAT-xz971.mzXML
Comment (hepcat72) [inline DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_glucose_mzxmls/BAT-xz971.mzXML:1]: C. Below this point, a bunch of test data was modified to add support for the new `mz_min` and `mz_max` fields/data.

File: DataRepo/management/commands/load_accucor_msruns.py
Comment (hepcat72) [inline DataRepo/management/commands/load_accucor_msruns.py:1]: C. Added options for the new default flags for low/high mz to the 2 command files below.

File: DataRepo/migrations/0030_auto_20240113_1215.py
Comment (hepcat72) [inline DataRepo/migrations/0030_auto_20240113_1215.py:1]: C. Added mz_min and mz_max fields and updated things like the string representation and the unique constraint.

File: DataRepo/tests/loading/test_lcms_metadata_loading.py
Comment (hepcat72) [inline DataRepo/tests/loading/test_lcms_metadata_loading.py:1]: C. There were a number of test updates. I'll comment on significant ones.

File: DataRepo/tests/loading/test_lcms_metadata_loading.py [hunk 6]
```diff
@@ -925,7 +927,11 @@ def test_PeakAnnotFileMismatches(self):
             )
         aes = ar.exception
         self.assertEqual(1, len(aes.exceptions))
-        self.assertEqual(PeakAnnotFileMismatches, type(aes.exceptions[0]))
+        self.assertEqual(
+            PeakAnnotFileMismatches,
+            type(aes.exceptions[0]),
+            msg=f"Exception should be PeakAnnotFileMismatches.  Got: [{type(aes.exceptions[0]).__name__}]",
+        )
```
Comment (hepcat72) [inline DataRepo/tests/loading/test_lcms_metadata_loading.py:934]: C. I added messages to a few tests for debugging purposes.

File: DataRepo/tests/loading/test_load_accucor_msruns.py [hunk 7]
```diff
@@ -36,7 +37,7 @@
 from DataRepo.utils.exceptions import (
     ConflictingValueErrors,
     DuplicatePeakGroup,
-    PolarityConflictErrors,
+    MzxmlConflictErrors,
```
Comment (hepcat72) [inline DataRepo/tests/loading/test_load_accucor_msruns.py:40]: C. This specific exception was converted to a generic conflict exception associated with data from an mzXML file.

File: DataRepo/tests/loading/test_load_accucor_msruns.py [hunk 8]
```diff
@@ -313,12 +315,18 @@ def test_duplicate_peak_group(self):
...
                 peak_group_attrs,
                 msrun_sample=peak_group.msrun_sample,
                 peak_annotation_file=peak_group.peak_annotation_file,
             )
+        exc = ar.exception
+        self.assertEqual(
+            DuplicatePeakGroup,
+            type(exc),
+            msg=f"{DuplicatePeakGroup} expected. Got [{type(exc).__name__}: {exc}].",
+        )
```
Comment (hepcat72) [inline DataRepo/tests/loading/test_load_accucor_msruns.py:329]: C. This is effectively the same. Just changed for debugging.

File: DataRepo/tests/loading/test_load_accucor_msruns.py [hunk 9]
```diff
@@ -437,9 +445,12 @@ def test_multiple_accucor_one_msrun(self):
         self.assertEqual(PeakData.objects.all().count(), PEAKDATA_ROWS * SAMPLES_COUNT)
 
     @tag("multi-msrun")
-    def test_duplicate_compounds_one_msrun(self):
```
Comment (hepcat72) [inline DataRepo/tests/loading/test_load_accucor_msruns.py:440]: C. The below test was converted to test for the new `AmbiguousMSRuns` exception. I realized this test wasn't checking for duplicate compounds anyway, and the conflict test was happening elsewhere in another separate test and there was no way to change the case to actually check for a `ConflictingValueErrors` exception without changing the data, and this file was used for other tests (I think).  Regardless, this test was redundant.

File: DataRepo/tests/loading/test_load_accucor_msruns.py [hunk 10]
```diff
@@ -459,14 +470,33 @@ def test_duplicate_compounds_one_msrun(self):
...
-        # 2 samples in the accucor file, so 2 PeakGroup peak annotation file conflicts
-        self.assertEqual(2, len(aes.exceptions[0].conflicting_value_errors))
+        self.assertTrue(isinstance(aes.exceptions[0], AmbiguousMSRuns))
 
-        # Check first file loaded
-        SAMPLES_COUNT = 2
-        PEAKDATA_ROWS = 7
-        MEASURED_COMPOUNDS_COUNT = 1  # Glucose
+    @tag("multi-msrun")
+    def test_resolve_ambiguous_msruns_error(self):
```
Comment (hepcat72) [inline DataRepo/tests/loading/test_load_accucor_msruns.py:476]: C. This was a new test I added to make sure that adding low/high mz values addresses the ambiguous msruns error.

File: DataRepo/utils/accucor_data_loader.py
Comment (hepcat72) [inline DataRepo/utils/accucor_data_loader.py:1]: C. The changes in this file should largely be straight-forward. Here's where the fun begins...

File: DataRepo/utils/accucor_data_loader.py [hunk 11]
```diff
@@ -1193,27 +1269,53 @@ def get_or_create_archive_file(
 
         mode = "rb" if is_binary else "r"
```
Comment (hepcat72) [inline DataRepo/utils/accucor_data_loader.py:1271]: C. Indented the code below to wrap in a try so I could prevent the script from exiting when conflicts are encountered in `ArchiveFile` record creation.

File: DataRepo/utils/accucor_data_loader.py [hunk 12]
```diff
@@ -1376,33 +1501,45 @@ def insert_peak_group(
                 existing_peak_group = PeakGroup.objects.get(
                     msrun_sample=msrun_sample, name=peak_group_attrs["name"]
                 )
-                conflicting_fields = []
-                existing_values = []
-                new_values = []
+                # TODO: Check more than just formula and peak_annotation_file.  Otherwise, if there are other
+                # differences, they will all be labeled inaccurately as AmbiguousMSRuns, though users should be able to
+                # figure it out.
```
Comment (hepcat72) [inline DataRepo/utils/accucor_data_loader.py:1506]: C. Note the TODO.  I should:

- [ ] TODO: Create an issue

**Comment by hepcat72** [comment]
BTW, @lparsons - I'm just about the start going over the review of 832. Just wanted to let you know that I have a branch of the Rabinowitz repo with all the necessary updates to work with this branch of the tracebase repo. I'll create a PR for the rabinowitz repo once the code is good to go.

File: DataRepo/utils/accucor_data_loader.py [hunk 13]
```diff
@@ -599,10 +653,21 @@ def fill_in_lcms_defaults(self):
...
                             continue
+                        # No default needed for mz_min or mz_max unless there are multiple MSRuns for the same sample
+                        # and polarity, which will be handled later
+                        # TODO: Fill in placeholders for mz_min and mz_max if the above scenario is encountered.
+                        if key == "mz_min":
+                            self.lcms_metadata[sample_header][key] = mz_min
+                            continue
+                        if key == "mz_max":
+                            self.lcms_metadata[sample_header][key] = mz_max
+                            continue
```
Comment (hepcat72) [inline DataRepo/utils/accucor_data_loader.py:669]: C. Regarding the mz_min and mz_max placeholder scenario... The placeholders should be added after data for an `AmbiguousMSRuns` exception has been buffered.  **That** is when the scenario is encountered.  And, the placeholders should not run into one another, e.g. if there are 3 mz ranges (and 3 accucor files for them).

- [ ] TODO: Create an issue for saving mz_min and mz_max placeholder values when data for an `AmbiguousMSRuns` exception is buffered.

File: DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_glucose_mzxmls/BAT-xz971.mzXML
Comment (lparsons) [inline DataRepo/data/tests/small_obob/small_obob_maven_6eaas_inf_glucose_mzxmls/BAT-xz971.mzXML:1]: I'm confused as to why `mzXML` files were modified in this PR. I understand that we are looking to differentiate between mzXML files produced by the same mass spec run in different scans, but I thought the `lowMz` and `highMz` were already present in the mzXML files.

**Review by lparsons — APPROVED** [review]
Nice work adding in the low/high mz values to the MSRun Sample. Assuming that the scan range and polarity differences are the only reasons for multiple mzXML files and multiple scans of the same compound (which seems likely), this should allow us to store one record per compound per scan, which is the goal.

Special attention was paid to allow for low/high mz values to be supplied when we don't have mzXML files and then checking to be sure supplied values are consistent if we do have mzXML files. This should allow us to handle a variety of cases and minimize the chances of errors, esp. for submissions with the mzXML files included.

Thanks for clarifying why the mzXML files were modified.

File: DataRepo/utils/accucor_data_loader.py (hunk 13 above)
Comment (hepcat72) [reply DataRepo/utils/accucor_data_loader.py:669]: #836

File: DataRepo/utils/accucor_data_loader.py (hunk 12 above)
Comment (hepcat72) [reply DataRepo/utils/accucor_data_loader.py:1506]: #837
